pkgs/applications/editors/pulsar/default.nix

```diff
@@ -0,0 +1,159 @@
+{ lib
+, stdenv
+, pkgs
+, fetchurl
+, wrapGAppsHook
+, glib
+, gtk3
+, atomEnv
+, xorg
+, libxkbcommon
+, hunspell
+, hunspellDicts
+, useHunspell ? true
+, languages ? [ "en_US" ]
+, nodePackages # We need ASAR to unpack the app.asar file and patch paths
+, python3 # Cursed patch to allow asar unpacking, as the builtin in the unpacked asar isn't there
+}:
+
+let
+  additionalLibs = lib.makeLibraryPath [
+    xorg.libxshmfence
+    libxkbcommon
+    xorg.libxkbfile
+  ];
+  newLibpath = "${atomEnv.libPath}:${additionalLibs}";
+  buildLocalePath = path: "searchPaths.push('${path}');";
+  localeDerivations = builtins.map (lang: hunspellDicts.${lang}) languages;
+  localePatchs = lib.concatMapStringsSep "" buildLocalePath localeDerivations;
+in
+stdenv.mkDerivation rec {
+  name = "pulsar";
+  version = "1.103.0";
+
+  src = fetchurl {
+    url = "https://github.com/pulsar-edit/pulsar/releases/download/v${version}/Linux.pulsar_${version}_amd64.deb ";
+    name = "${name}.deb";
+    sha256 = "16k3j9rw0mshv2gfhwrccpn2d2704whw640qjgzwkal0lwjpx49x";
+  };
+
+  nativeBuildInputs = [
+    wrapGAppsHook # Fix error: GLib-GIO-ERROR **: No GSettings schemas are installed on the system
+  ];
+
+  buildInputs = [
+    gtk3 # Fix error: GLib-GIO-ERROR **: Settings schema 'org.gtk.Settings.FileChooser' is not installed
+    xorg.libxkbfile # Fix error on electron: libxkbfile.so.1: cannot open shared object file: No such file or directory
+    nodePackages.asar # We need ASAR to unpack the app.asar file and patch paths
+  ];
+
+  dontBuild = true;
+  dontConfigure = true;
+
+  unpackPhase = ''
+    ar p $src data.tar.xz | tar xJ ./usr/ ./opt/
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out
+    mv opt usr $out
+
+    runHook postInstall
+  '';
+
+  preFixup = ''
+    opt=$out/opt/Pulsar
+
+    patchHunspell=${if useHunspell then "1" else "0"}
+
+    # Patch the hunspell dictionaries
+    if [ $patchHunspell -eq 1 ]; then
+      # We need to patch the already existing app.asar.unpacked, and add python3
+      # to resources/app.asar.unpacked/node_modules/tree-sitter-bash/build/node_gyp_bins/python3
+      rm $opt/resources/app.asar.unpacked/node_modules/tree-sitter-bash/build/node_gyp_bins/python3
+      ln -s ${python3}/bin/python3 $opt/resources/app.asar.unpacked/node_modules/tree-sitter-bash/build/node_gyp_bins/python3
+
+      # We need to extract the app.asar file
+      asar extract $opt/resources/app.asar ./app
+
+      # Remove lines 114 to 116 (inclusive) and replace with ''${localePaths}
+      sed -i '114,116d' ./app/node_modules/spell-check/lib/locale-checker.js
+      echo 'sed -i "114i ${localePatchs}" ./app/node_modules/spell-check/lib/locale-checker.js'
+      sed -i "114i ${localePatchs}" ./app/node_modules/spell-check/lib/locale-checker.js
```

Comment:
This seems abit fragile, any line change could break the patching and only cause an error when this code gets run after the build.

Haven't taken a look at the exact code this replaces and not too fluent in JS, but maybe you could instead write a patch with smth like
```js
const NIX_LOCALE_PATHS = [ @localePaths@ ];
for (let nixLocalePathsIndex = 0; nixLocalePathsIndex < NIX_LOCALE_PATHS.length; ++nixLocalePathsIndex) {
  searchPaths.push(NIX_LOCALE_PATHS[nixLocalePathsIndex]);
}
```
and use [`substituteAll`](https://nixos.org/manual/nixpkgs/stable/#fun-substituteAll) on it to fill in the placeholder?
```nix
localePaths = lib.strings.concatMapStringsSep ", " (path: "'${path}'") localeDerivations;
```

Reply:
The code roughly looks like this upstream:

```javascript
        // Add operating system specific paths to the search list.
        if (this.checkDefaultPaths) {
            if (env.isLinux()) {
                searchPaths.push('/usr/share/hunspell');
                searchPaths.push('/usr/share/myspell');
                searchPaths.push('/usr/share/myspell/dicts');
            }

            if (env.isDarwin()) {
                searchPaths.push('/');
                searchPaths.push('/System/Library/Spelling');
            }

            if (env.isWindows()) {
                searchPaths.push('C:\\');
            }
        }
```

The idea was to remove the pathes in the `if (env.isLinux()) {` block, and instead replace them with our own paths.

Is it necessary to remove those path ? It would enhance compatibility on non NixOS Linux systems, but it would reference something not managed by Nix.

If it's not necessary, I can just `sed "if (this.checkDefaultPaths) {"i ${localePatchs}`, to rely on the actual code instead of line numbers